### PR TITLE
Telemetry Command

### PR DESCRIPTION
**Pull request by lramos15**
Adds a `--telemetry` command to the VS Code CLI which dumps all the the telemetry events which VS Code collects.

### CODE REVIEW

**Comment by sbatten** [comment]
@joaomoreno please review the build pipeline changes.

File: build/azure-pipelines/linux/product-build-linux-alpine.yml [hunk 1]
```diff
@@ -64,7 +64,7 @@ steps:
 - script: |
     set -e
     ./build/azure-pipelines/linux/prebuild-alpine.sh
-  displayName: Prepare build
+  displayName: Prepare builds
```
Comment (sbatten) [inline build/azure-pipelines/linux/product-build-linux-alpine.yml]: unintentional?

File: src/vs/platform/environment/node/argv.ts [hunk 2]
```diff
@@ -217,6 +220,39 @@ export function buildVersionMessage(version: string | undefined, commit: string
...
+	const mergeTelemetry = (contents: string, dirName: string) => {
+		const telemetryData = JSON.parse(contents);
+		mergedTelemetry[dirName] = telemetryData;
+	};
+	telemetryJsonFolders.forEach((folder) => {
+		const contents = readFileSync(join(extensionsPath, folder, 'telemetry.json')).toString();
+		mergeTelemetry(contents, folder);
+	});
+	try {
+		let contents = readFileSync(path.join(appRoot, 'out/', 'telemetry-core.json')).toString();
```
Comment (sbatten) [inline src/vs/platform/environment/node/argv.ts]: don't use path from node. use our implementation of join which is already imported into this file

File: src/vs/platform/environment/node/argv.ts [hunk 3]
```diff
@@ -217,6 +220,39 @@ export function buildVersionMessage(version: string | undefined, commit: string
 	return `${version || localize('unknownVersion', "Unknown version")}\n${commit || localize('unknownCommit', "Unknown commit")}\n${process.arch}`;
 }
 
+export function buildTelemetryMessage(appRoot: string, extensionsPath: string): string {
+	// const contents = readFileSync('./telemetry.json');
```
Comment (sbatten) [inline src/vs/platform/environment/node/argv.ts]: old comment

File: src/vs/platform/environment/node/argv.ts [hunk 4]
```diff
@@ -217,6 +220,39 @@ export function buildVersionMessage(version: string | undefined, commit: string
...
+	// Simple function to merge the telemetry into one json object
+	const mergeTelemetry = (contents: string, dirName: string) => {
+		const telemetryData = JSON.parse(contents);
+		mergedTelemetry[dirName] = telemetryData;
+	};
+	telemetryJsonFolders.forEach((folder) => {
+		const contents = readFileSync(join(extensionsPath, folder, 'telemetry.json')).toString();
+		mergeTelemetry(contents, folder);
+	});
+	try {
```
Comment (sbatten) [inline src/vs/platform/environment/node/argv.ts]: we try-catch core but not extension dir?

**Review by sbatten — CHANGES_REQUESTED** [review]

**Review by sbatten — APPROVED** [review]

**Review by Tyriar — COMMENTED** [review]
Can you reduce duplication by deferring to script(s) in `./build/azure-pipelines`?

**Review by Tyriar — APPROVED** [review]
LGTM, feel free to defer the code duplication if you're sure it's good.

**Comment by lramos15** [comment]
Deferring to scripts is good but I won't have adequate testing in time for it to get into the June milestone, so I will look into it for the July one as I know this works and don't want to break anything

**Comment by joaomoreno** [comment]
@lramos15 @Tyriar @sbatten I have a few questions:

1. Why does this need to be run per platform?
2. If this works by parsing GDPR formatted contents, why do we need to run Code at all to spit out information? Why isn't this just done statically against our sources?
3. If `code --telemetry` spits out all the information, what's the purpose of https://github.com/microsoft/vscode-telemetry-extractor?
4. Why are we cloning https://github.com/microsoft/vscode-telemetry-extractor instead of publishing that as a package to npm and using a specific version of that in our `devDependencies`? We want to always have reproducible builds... and currently builds will always get `master` from that repo, voiding this contract.

File: build/azure-pipelines/darwin/product-build-darwin.yml [hunk 5]
```diff
@@ -59,6 +59,21 @@ steps:
...
+    git clone https://github.com/microsoft/vscode-telemetry-extractor.git
+    cd vscode-telemetry-extractor
+    npm install
+    ./node_modules/typescript/bin/tsc
+    npm run setup-extension-repos
+    node ./out/cli-extract.js --sourceDir ../s --excludedDirPattern extensions  --outputDir . --applyEndpoints --includeIsMeasurement --patchWebsiteEvents
+    node ./out/cli-extract-extensions.js --sourceDir ./src/telemetry-sources/ --outputDir . --applyEndpoints --includeIsMeasurement
+    mv declarations-resolved.json ../s/src/telemetry-core.json
+    mv declarations-extensions-resolved.json ../s/src/telemetry-extensions.json
+    echo 'Moved Files'
```
Comment (joaomoreno) [inline build/azure-pipelines/darwin/product-build-darwin.yml]: This should just be removed.

File: src/vs/platform/environment/node/argv.ts [hunk 6]
```diff
@@ -217,6 +219,38 @@ export function buildVersionMessage(version: string | undefined, commit: string
...
+			const contents = readFileSync(join(extensionsPath, folder, 'telemetry.json')).toString();
+			mergeTelemetry(contents, folder);
+		});
+		let contents = readFileSync(join(appRoot, 'out/', 'telemetry-core.json')).toString();
+		mergeTelemetry(contents, 'vscode-core');
+		contents = readFileSync(join(appRoot, 'out/', 'telemetry-extensions.json')).toString();
+		mergeTelemetry(contents, 'vscode-extensions');
+		return JSON.stringify(mergedTelemetry, null, 4);
+	} catch (err) {
+		return 'Unable to read VS Code telemetry events!';
```
Comment (joaomoreno) [inline src/vs/platform/environment/node/argv.ts]: Just remove the whole `try catch` block and let the error bubble up and be printed to the console.

File: src/vs/platform/environment/node/argv.ts [hunk 7]
```diff
@@ -217,6 +219,38 @@ export function buildVersionMessage(version: string | undefined, commit: string
 	return `${version || localize('unknownVersion', "Unknown version")}\n${commit || localize('unknownCommit', "Unknown commit")}\n${process.arch}`;
 }
 
+export function buildTelemetryMessage(appRoot: string, extensionsPath: string): string {
+	try {
```
Comment (joaomoreno) [inline src/vs/platform/environment/node/argv.ts]: This needs to be an `async` function. Every single `*Sync` call here should be an `await *` call.

File: build/azure-pipelines/linux/product-build-linux.yml [hunk 8]
```diff
@@ -60,6 +60,21 @@ steps:
...
+    set -e
+    cd ..
+    git clone https://github.com/microsoft/vscode-telemetry-extractor.git
+    cd vscode-telemetry-extractor
+    npm install
+    ./node_modules/typescript/bin/tsc
+    npm run setup-extension-repos
+    node ./out/cli-extract.js --sourceDir ../s --excludedDirPattern extensions  --outputDir . --applyEndpoints --includeIsMeasurement --patchWebsiteEvents
+    node ./out/cli-extract-extensions.js --sourceDir ./src/telemetry-sources/ --outputDir . --applyEndpoints --includeIsMeasurement
+    mv declarations-resolved.json ../s/src/telemetry-core.json
```
Comment (joaomoreno) [inline build/azure-pipelines/linux/product-build-linux.yml]: I know I suggested to put the files in `src`, but not knowing the context then I didn't know better. These files should live next to `package.json` in the built product, right here:

image

In order to achieve this:

1. Move them into `$BUILD_STAGINGDIRECTORY/telemetry*.json` instead
2. If `process.env['BUILD_STAGINGDIRECTORY']` is set, have a `gulp.src('telemetry*.json', { cwd: process.env['BUILD_STAGINGDIRECTORY'] })` stream piping into the pipeline mentioned above.

**Comment by joaomoreno** [comment]
I've talked to @kieferrm about this and he helped me understand this better. Answering my own questions:

1. This is not platform specific. If we go along with https://github.com/microsoft/vscode/issues/75983 this won't be a problem.
2. I got the flow wrong. This is statically computed. Code spitting out telemetry IS the feature this PR implements.
3. Understood.

I'd still **highly** recommend publishing https://github.com/microsoft/vscode-telemetry-extractor to NPM and locking it in our yarn.lock file.

File: build/azure-pipelines/darwin/product-build-darwin.yml (hunk 5 above)
Comment (lramos15) [reply build/azure-pipelines/darwin/product-build-darwin.yml]: So Darwin doesn't need a specific build script?

File: build/azure-pipelines/linux/product-build-linux.yml (hunk 8 above)
Comment (lramos15) [reply build/azure-pipelines/linux/product-build-linux.yml]: Okay sounds good, I'll look into this

**Comment by lramos15** [comment]
Yeah sorry for not explaining it well. Ideally we would only compute this once and copy it somehow to all the platforms, but since each platform is run in parallel I wasn't sure how to accomplish that.  You are correct that it should be an NPM package, but the telemetry extractor is not open source yet so the plan was to hold off on releasing an NPM package until the tool was open sourced.

**Comment by joaomoreno** [comment]
> You are correct that it should be an NPM package, but the telemetry extractor is not open source yet so the plan was to hold off on releasing an NPM package until the tool was open sourced.

Then let's lock the clone to a specific commit ID for now.

File: src/vs/platform/environment/node/argv.ts (hunk 7 above)
Comment (lramos15) [reply src/vs/platform/environment/node/argv.ts]: Is there a reason this should be async? I was just kind of modeling off of how we build the help message and that isn't async so I figured this didn't need to be either as the user is going to want for the command to finish anyways.

Comment (joaomoreno) [reply src/vs/platform/environment/node/argv.ts]: Oh yeah, totally right, sorry about that!

File: build/azure-pipelines/darwin/product-build-darwin.yml (hunk 5 above)
Comment (joaomoreno) [reply build/azure-pipelines/darwin/product-build-darwin.yml]: I mean the `echo`. :+1:

File: src/vs/platform/environment/common/environment.ts [hunk 9]
```diff
@@ -13,6 +13,7 @@ export interface ParsedArgs {
 	_urls?: string[];
 	help?: boolean;
 	version?: boolean;
+	telemetry?: boolean;
```
Comment (sbatten) [inline src/vs/platform/environment/common/environment.ts]: @lramos15 shouldn't break anything but missed this arg when removing the command

File: build/gulpfile.vscode.js [hunk 10]
```diff
@@ -62,6 +62,8 @@ const vscodeResources = [
 	'out-build/bootstrap-amd.js',
 	'out-build/bootstrap-window.js',
 	'out-build/paths.js',
+	'out-build/telemetry-core.json',
+	'out-build/telemetry-extensions.json',
```
Comment (sbatten) [inline build/gulpfile.vscode.js]: @lramos15 I noticed this code isn't in master anymore but it wasn't removed during the PR to remove the command. What happened to it?

**Pull request merged**